File: hillClimbing.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <cmath>
#include <limits>

double objectiveFunction(double x) {
    return x * x;
}

double getRandom(double lower_bound, double upper_bound) {
    return lower_bound + static_cast<double>(rand()) / RAND_MAX * (upper_bound - lower_bound);
}

double hillClimbing(double lower_bound, double upper_bound, double step_size, int max_iterations) {
    // Initialize random seed
    srand(static_cast<unsigned>(time(0)));

    // Step 1: Initialize the current solution
    double current_solution = getRandom(lower_bound, upper_bound);
    double current_value = objectiveFunction(current_solution);

    for (int iteration = 0; iteration < max_iterations; ++iteration) {
        // Step 4: Generate neighboring solution by perturbing the current solution
        double new_solution = current_solution + getRandom(-step_size, step_size);
        
        // Keep new solution within bounds
        if (new_solution < lower_bound) new_solution = lower_bound;
        if (new_solution > upper_bound) new_solution = upper_bound;

        double new_value = objectiveFunction(new_solution);

        // Update if the new solution is better
        if (new_value < current_value) {
            current_solution = new_solution;
            current_value = new_value;
        } else {
            // Decrease step size to refine the search
            step_size *= 0.9;
        }
    }

    return current_solution;
}

int main() {
    double lower_bound = -5.0;
    double upper_bound = 5.0;
    double step_size = 1.0;
    int max_iterations = 1000;

    // Run hill climbing
    double solution = hillClimbing(lower_bound, upper_bound, step_size, max_iterations);

    std::cout << "Optimal solution: x = " << solution << std::endl;
    std::cout << "Minimum value of f(x) = " << objectiveFunction(solution) << std::endl;

    return 0;
}

File: hillClimbing.py
import numpy as np

# Objective function: f(x) = x^2
def objective(x):
    return x**2

# Hill climbing algorithm
def hillclimbing(objective, bounds, n_iterations, step_size):
    # Initialize the current solution randomly within bounds
    solution = np.random.uniform(bounds[0], bounds[1])
    # Evaluate the current solution
    solution_eval = objective(solution)
    print(f"Initial solution: x = {solution}, f(x) = {solution_eval}")
    
    # Run the hill climbing search
    for i in range(n_iterations):
        # Generate a new candidate by perturbing the current solution
        candidate = solution + np.random.randn() * step_size
        # Ensure the candidate is within the bounds
        candidate = np.clip(candidate, bounds[0], bounds[1])
        # Evaluate the candidate solution
        candidate_eval = objective(candidate)
        
        # If the candidate is better, update the solution
        if candidate_eval < solution_eval:
            solution, solution_eval = candidate, candidate_eval
            print(f"Iteration {i+1}: x = {solution}, f(x) = {solution_eval}")
    
    return solution, solution_eval

# Seed the random number generator for reproducibility
np.random.seed(5)

# Define parameters
bounds = [-5.0, 5.0]  # Search space bounds
n_iterations = 100     # Number of iterations
step_size = 0.1        # Maximum step size

# Perform the hill climbing search
best_solution, best_score = hillclimbing(objective, bounds, n_iterations, step_size)

# Print the final result
print("Done!")
print(f"Best solution: x = {best_solution}, f(x) = {best_score}")
